src/java/htsjdk/samtools/SAMFileWriterFactory.java

```diff
@@ -311,28 +332,91 @@ public CRAMFileWriter makeCRAMWriter(final SAMFileHeader header, final OutputStr
         return writer;
     }
 
+    /**
+     * Create a CRAMFileWriter on an output file. Requires input record to be presorted to match the
+     * sort order defined by the input header.
+     *
+     * Note: does not honor factory settings for USE_ASYNC_IO.
+     * Read names and tags are preserved.
+     *
+     * @param header entire header. Sort order is determined by the sortOrder property of this arg.
+     * @param outputFile where to write the output.  Must end with .sam, .bam or .cram.
+     * @param referenceFasta reference sequence file
+     * @return CRAMFileWriter
+     *
+     */
     public CRAMFileWriter makeCRAMWriter(final SAMFileHeader header, final File outputFile, final File referenceFasta) {
+        return createCRAMWriter(header, true, outputFile, referenceFasta);
+    }
 
-        final boolean createIndex = this.createIndex && IOUtil.isRegularPath(outputFile);
-        if (this.createIndex && !createIndex) {
-            System.err.println("Cannot create index for CAM because output file is not a regular file: " + outputFile.getAbsolutePath());
-        }
 
-        try {
+    /**
+     * Create a CRAMFileWriter on an output file.
+     *
+     * Note: does not honor factory setting for USE_ASYNC_IO.
+     * Read names and tags are preserved.
+     *
+     * @param header entire header. Sort order is determined by the sortOrder property of this arg.
+     * @param presorted  if true, SAMRecords must be added to the SAMFileWriter in order that agrees with header.sortOrder.
+     * @param outputFile where to write the output.  Must end with .sam, .bam or .cram.
+     * @param referenceFasta reference sequence file
+     * @return CRAMFileWriter
+     *
+     */
+    public CRAMFileWriter makeCRAMWriter(final SAMFileHeader header, final boolean presorted, final File outputFile, final File referenceFasta) {
+        return createCRAMWriter(header, presorted, outputFile, referenceFasta);
+    }
+
+    /**
+     * Create a CRAMFileWriter on an output file.
+     *
+     * Note: does not honor factory setting for USE_ASYNC_IO.
+     *
+     * @param header entire header. Sort order is determined by the sortOrder property of this arg.
+     * @param presorted  if true, SAMRecords must be added to the SAMFileWriter in order that agrees with header.sortOrder.
+     * @param outputFile where to write the output.  Must end with .sam, .bam or .cram.
+     * @param referenceFasta reference sequence file
+     * @return CRAMFileWriter
+     *
+     * Read names and tags are preserved.
+     */
+    private CRAMFileWriter createCRAMWriter(final SAMFileHeader header, final boolean presorted, final File outputFile, final File referenceFasta) {
```

Comment:
It is confusing; the problem is the first (3 arg) makeCRAMWriter method, which doesn't take a presorted flag and takes a stream instead of a file (so it can't and doesn't honor the other factory settings). It creates the CRAMFileWriter directly. If it wasn't already public I would lobby to remove it since I think it has limited use.

The other two overloads call createCRAMWriter who's main job is to propagate the factory settings, but it depends on being passed a file (to create parallel index and md5 files). I'm going to rename createCRAMWriter to reflect that and factor out the 2 duplicated lines of initialization code. Other than that I don't see a refactoring that doesn't add even more confusion, but if you have a suggestion do let me know.